active-rfcs/0000-function-api.md

```diff
@@ -0,0 +1,1176 @@
+
+``` js
+export default {
+  setup() {
+    // stopped automatically when the component unmounts
+    watch(/* ... */)
+  }
+}
+```
+
+### Effect Cleanup
+
+Sometimes the watcher callback will perform async side effects that need to be invalidated when the watched value changes. The watcher callback receives a 3rd argument that can be used to register a cleanup function. The cleanup function is called when:
+
+- the watcher is about to re-run
+- the watcher is stopped (i.e. when the component is unmounted if `watch` is used inside `setup()`)
+
+``` js
+watch(idValue, (id, oldId, onCleanup) => {
+  const token = performAsyncOperation(id)
+  onCleanup(() => {
+    // id has changed or watcher is stopped.
+    // invalidate previously pending async operation
+    token.cancel()
+  })
+})
+```
+
+We are registering cleanup via a passed-in function instead of returning it from the callback (like React `useEffect`) because the return value is important for async error handling. It is very common for the watcher callback to be an async function when performing data fetching:
+
+``` js
+const data = value(null)
+watch(getId, async (id) => {
+  data.value = await fetchData(id)
+})
+```
+
+An async function implicitly returns a Promise, but the cleanup function needs to be registered immediately before the Promise resolves. In addition, Vue relies on the returned Promise to automatically handle potential errors in the Promise chain.
+
+### Watcher Callback Timing
+
+By default, all watcher callbacks are fired **after current renderer flush.** This ensures that when callbacks are fired, the DOM will be in already-updated state. If you want a watcher callback to fire before flush or synchronously, you can use the `flush` option:
+
+``` js
+watch(
+  () => count.value + 1,
+  () => console.log(`count changed`),
+  {
+    flush: 'post', // default, fire after renderer flush
+    flush: 'pre', // fire right before renderer flush
+    flush: 'sync' // fire synchronously
+  }
+)
+```
+
+### Full `watch` Options
+
+``` ts
+interface WatchOptions {
+  lazy?: boolean
+  deep?: boolean
+  flush?: 'pre' | 'post' | 'sync'
+  onTrack?: (e: DebuggerEvent) => void
+  onTrigger?: (e: DebuggerEvent) => void
+}
+
+interface DebuggerEvent {
+  effect: ReactiveEffect
+  target: any
+  key: string | symbol | undefined
+  type: 'set' | 'add' | 'delete' | 'clear' | 'get' | 'has' | 'iterate'
+}
+```
+
+- `lazy` is the opposite of 2.x's `immediate` option.
+- `deep` works the same as 2.x
+- `onTrack` and `onTrigger` are hooks that will be called when a dependency is tracked or the watcher getter is triggered. They receive a debugger event that contains information on the operation that caused the track / trigger.
+
+## Lifecycle Hooks
+
+All current lifecycle hooks will have an equivalent `onXXX` function that can be used inside `setup()`:
+
+``` js
+import { onMounted, onUpdated, onUnmounted } from 'vue'
```

Comment:
I understand the comfort of just prefixing the existing lifecycle hooks with "on". However, I think that those names really should use present tense (`onMount`, `onUpdate` etc.) if there are no strong objections.

This would match the API of the web platform (`onload`, `onclick`, ...) as well as other Frameworks ([Angular](https://angular.io/guide/lifecycle-hooks), [Svelte](https://svelte.dev/docs#onMount)).

Reply:
`onMount` doesn't convey wether it runs before or after mounting. It's only clear by knowing that there is also a `beforeMount` hook.

OTOH, `onMounted`, makes it clear that the component  already has mounted before this hook is being run.

Reply:
Yes, that's the one benefit I saw in the naming. I still find the difference to established conventions a little bit weird. FWIW I mostly see `before/afterXXX` when past tense is used. I'd assume that this is because `on` grammatically doesn't match with past tense.

However, like many of us, I'm not a native English speaker. Would be good to hear from some of them how this plays out in the language.

Reply:
i dont think this is a good idea in general. i am not that sure about when exactly each hook is triggered, but if a hook is called `onUpdate`, idk if the update is about to happen, happens right now or has happened already, while `onUpdated` clearly communicates that the update already happened and one would expect all the work the framework does during an update is done.

Reply:
I don't really understand the arguments that the tense is confusing. The correct terminology would be `onMount`. If it happened before the component was mounted then, well, it wouldn't be `onMount`. I never expect an `onclick` event to happen before click 🤔

The consistency amongst current web standards and other frameworks is above be ultra clear about a case that doesn't make sense anyways.